### ISSUE TITLE
Wrong metrics for postgres (apps.plugin)

### ISSUE THREAD

**Issue by ossjet** [firehol/netdata#3244]
Hi there.

I've noticed that postgres' RSS values in top are differs from what netdata shows. For example:

```
# ps -e | grep postgres
 2449 ?        00:19:14 postgres
 6428 ?        00:01:37 postgres
11117 ?        00:06:26 postgres
19827 ?        07:26:24 postgres
19930 ?        00:00:49 postgres
19936 ?        00:09:15 postgres
19937 ?        00:10:13 postgres
19938 ?        00:06:03 postgres
19940 ?        00:00:23 postgres
19941 ?        01:33:07 postgres
26666 ?        00:16:28 postgres

# cat /proc/19936/statm
448935 403934 403681 1612 0 625 0

# grep -E '^Vm' /proc/19936/status
VmPeak:  1796184 kB
VmSize:  1795740 kB
VmLck:         0 kB
VmPin:         0 kB
VmHWM:   1617972 kB
VmRSS:   1615736 kB
VmData:     2368 kB
VmStk:       132 kB
VmExe:      6448 kB
VmLib:     45912 kB
VmPTE:      3424 kB
VmPMD:        20 kB
VmSwap:     1700 kB
```
As I understood from netdata source code [(apps_plugin.c)](https://github.com/firehol/netdata/blob/master/src/apps_plugin.c#L1076)  you're reading the second field from the /proc/PID/statm file. 

From man proc:

```
...
       /proc/[pid]/statm
              Provides information about memory usage, measured in pages.  The columns are:

                  size       (1) total program size
                             (same as VmSize in /proc/[pid]/status)
                  resident   (2) resident set size
                             (same as VmRSS in /proc/[pid]/status)
                  share      (3) shared pages (i.e., backed by a file)
...
```

As I showed in the beginning, just one of postgres procs uses more than 1.5 GB:

`403934 (memory pages count) * 4096 (page size)  / 1024 / 1024 ~= 1.5GB`

But, here's what netdata shows:

```
# curl -s http://localhost:19999/api/v1/allmetrics | grep -E 'NETDATA_APPS_.+_POSTGRES'
NETDATA_APPS_PIPES_POSTGRES="12"      # open pipes
NETDATA_APPS_SOCKETS_POSTGRES="8"      # open sockets
NETDATA_APPS_FILES_POSTGRES="411"      # open files
NETDATA_APPS_LWRITES_POSTGRES="44"      # kilobytes/s
NETDATA_APPS_LREADS_POSTGRES="1307"      # kilobytes/s
NETDATA_APPS_PWRITES_POSTGRES="32"      # kilobytes/s
NETDATA_APPS_PREADS_POSTGRES="32"      # kilobytes/s
NETDATA_APPS_MINOR_FAULTS_POSTGRES="24968"      # page faults/s
NETDATA_APPS_MAJOR_FAULTS_POSTGRES="0"      # page faults/s
NETDATA_APPS_CPU_SYSTEM_POSTGRES="9"      # cpu time %
NETDATA_APPS_CPU_USER_POSTGRES="18"      # cpu time %
NETDATA_APPS_PROCESSES_POSTGRES="11"      # processes
NETDATA_APPS_THREADS_POSTGRES="11"      # threads
NETDATA_APPS_VMEM_POSTGRES="16110"      # MB
NETDATA_APPS_MEM_POSTGRES="22"      # MB
NETDATA_APPS_CPU_POSTGRES="27"      # cpu time %
```
As you can see netdata found all 11 process of postgres, but it still reports 22 MB of resident memory (RSS). 

Is it a bug or I missed something?

**Comment by ktsaou** [comment]
Hi,

apps.plugin excludes shared memory:

![image](https://user-images.githubusercontent.com/2662304/34918903-8ae5e1da-f963-11e7-932c-ba96692a34cb.png)

The reason is that because it groups / aggregates processes, without excluding shared memory it would report a multiple of the collected value.

To properly report shared memory, apps.plugin would need to read all the shared memory mappings of all processes to see which regions overlap.

So, this is by design and yes, it is a side effect / limitation due to the design decision made to aggregate processes.

**Comment by n0guest** [comment]
> The reason is that because it groups / aggregates processes, without excluding shared memory it would report a multiple of the collected value.

You mean 2nd field from statm minus 3rd field from the same file? I.e for:

```
# cat /proc/28298/statm
466753 415764 399923 1612 0 17288 0
```

It would be (415764-399923)*4096/1024/1024 ~= 61Mb of RSS without "inter-process shared memory"?

> To properly report shared memory, apps.plugin would need to read all the shared memory mappings of all processes to see which regions overlap.

And i guess it would costs too much of resources (cost of computing this every seconds)?

PS: Main reason of this question is very simple. Tools like top/atop/etc are showing the rss/virt/shared fields for memory metrics. I know, that computing a real memory consumption in Linux is very tricky thing (shared memory, shared libs, KSM, etc). But many users are "used to" values from top/atop. And different values in Netdata just causing misunderstandings and follow-up questions.

**Comment by ktsaou** [comment]
> You mean 2nd field from statm minus 3rd field from the same file?

yes. This is the point these values are read: https://github.com/ktsaou/netdata/blob/34bae08bb354bbd7977bca3b63aeb1fb9bf58494/src/apps_plugin.c#L1092-L1093

Then, they are aggregated for the process group defined in `apps_groups.conf`: https://github.com/ktsaou/netdata/blob/34bae08bb354bbd7977bca3b63aeb1fb9bf58494/src/apps_plugin.c#L2543-L2544

Then, the aggregated difference of the 2 metrics is sent to netdata: https://github.com/ktsaou/netdata/blob/34bae08bb354bbd7977bca3b63aeb1fb9bf58494/src/apps_plugin.c#L2981

> And i guess it would costs too much of resources (cost of computing this every seconds)?

Haven't tried it, but I think it would be quite heavy to do for all processes every second.

> But many users are "used to" values from top/atop. And different values in Netdata just causing misunderstandings and follow-up questions.

Yes you are right. I have made the most to be as close as possible.  But each tool has its strengths and weaknesses.

For example, it was relatively simple for netdata to also report the cpu usage of the exited children processes, while none of the console `top` like tools does. Check it here: https://github.com/firehol/netdata/wiki/Linux-console-tools%2C-fail-to-report-per-process-CPU-usage-properly

But it is almost impossible to report shared memory properly. Probably we need another netdata plugin, optimized for just shared memory reporting...

**Comment by n0guest** [comment]
> But it is almost impossible to report shared memory properly. Probably we need another netdata plugin, optimized for just shared memory reporting...

It's not about proper values, it's mostly about having at least "rough" numbers of memory consumption and deviation from "normal" behaviour. Without "shared memory" we have to use top/atop to get current consumption, which eventually breaks our "usage case" (Netdata as local "short range" metrics and export to Prometheus as longer one).

Is it possible to have at least "inaccurate" metrics in Netdata for shared memory? If not, we'll try to solve it by ourselfs (may be via custom python.d plugin).

**Comment by ktsaou** [comment]
Unfortunately there is no solution out of the box now.
So, I guess you will have to write a plugin for it.
